File: services/terrapod/api/routers/execution_hooks.py
"""Execution hook library CRUD + workspace association (#619).

Terrapod-native management surface at ``/api/terrapod/v1/execution-hooks``.
A hook is a reusable custom-shell step run inside the runner Job at one of five
fixed points; it reaches a workspace only via explicit association (no global
flag). Managing hooks is platform-admin-gated (like variable sets) because a
hook is operator-supplied code that runs with the runner's cloud identity.

Consumed by ``web/src/app/admin/execution-hooks/*`` and go-terrapod (PR2).
"""

import uuid
from datetime import UTC

from fastapi import APIRouter, Body, Depends, HTTPException, Path
from fastapi.responses import JSONResponse
from sqlalchemy import inspect as sa_inspect
from sqlalchemy import select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from terrapod.api.dependencies import AuthenticatedUser, require_admin
from terrapod.db.models import ExecutionHook, ExecutionHookWorkspace, Workspace
from terrapod.db.session import get_db
from terrapod.services import execution_hook_service

router = APIRouter(tags=["execution-hooks"])


def _rfc3339(dt) -> str:
    if dt is None:
        return ""
    return dt.astimezone(UTC).strftime("%Y-%m-%dT%H:%M:%SZ")


def _hook_json(h: ExecutionHook) -> dict:
    """Serialize an ExecutionHook to JSON:API. Guards relationship access with
    ``sa_inspect(...).dict`` so it never triggers a lazy load when the
    assignments weren't eagerly loaded."""
    loaded = sa_inspect(h).dict
    assignments = h.workspace_assignments if "workspace_assignments" in loaded else []
    relationships: dict = {
        "workspaces": {
            "data": [{"id": f"ws-{a.workspace_id}", "type": "workspaces"} for a in assignments]
        }
    }
    return {
        "id": f"hook-{h.id}",
        "type": "execution-hooks",
        "attributes": {
            "name": h.name,
            "description": h.description,
            "hook-point": h.hook_point,
            "script": h.script,
            "enabled": h.enabled,
            "priority": h.priority,
            "workspace-count": len(assignments),
            "created-at": _rfc3339(h.created_at),
            "updated-at": _rfc3339(h.updated_at),
        },
        "relationships": relationships,
    }


async def _get_hook(hook_id: str, db: AsyncSession) -> ExecutionHook:
    try:
        h_uuid = uuid.UUID(hook_id.removeprefix("hook-"))
    except ValueError as exc:
        raise HTTPException(status_code=404, detail="Execution hook not found") from exc
    result = await db.execute(
        select(ExecutionHook)
        .where(ExecutionHook.id == h_uuid)
        .options(selectinload(ExecutionHook.workspace_assignments))
    )
    hook = result.scalar_one_or_none()
    if hook is None:
        raise HTTPException(status_code=404, detail="Execution hook not found")
    return hook


@router.get("/execution-hooks")
async def list_hooks(
    user: AuthenticatedUser = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """List all execution hooks. Requires admin."""
    result = await db.execute(
        select(ExecutionHook)
        .options(selectinload(ExecutionHook.workspace_assignments))
        .order_by(ExecutionHook.name)
    )
    hooks = result.scalars().all()
    return JSONResponse(content={"data": [_hook_json(h) for h in hooks]})


@router.post("/execution-hooks", status_code=201)
async def create_hook(
    body: dict = Body(...),
    user: AuthenticatedUser = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """Create an execution hook. Requires admin."""
    attrs = body.get("data", {}).get("attributes", {})
    name = (attrs.get("name") or "").strip()
    if not name:
        raise HTTPException(status_code=422, detail="Execution hook name is required")
    hook_point = attrs.get("hook-point", "")
    execution_hook_service.validate_hook_point(hook_point)

    hook = ExecutionHook(
        name=name,
        description=attrs.get("description", ""),
        hook_point=hook_point,
        script=attrs.get("script", ""),
        enabled=attrs.get("enabled", True),
        priority=int(attrs.get("priority", 0) or 0),
    )
    db.add(hook)
    try:
        await db.commit()
    except IntegrityError as exc:
        await db.rollback()
        raise HTTPException(
            status_code=409, detail=f"An execution hook named '{name}' already exists"
        ) from exc
    await db.refresh(hook, ["workspace_assignments"])
    return JSONResponse(content={"data": _hook_json(hook)}, status_code=201)


@router.get("/execution-hooks/{hook_id}")
async def show_hook(
    hook_id: str = Path(...),
    user: AuthenticatedUser = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """Show an execution hook. Requires admin."""
    hook = await _get_hook(hook_id, db)
    return JSONResponse(content={"data": _hook_json(hook)})


@router.patch("/execution-hooks/{hook_id}")
async def update_hook(
    hook_id: str = Path(...),
    body: dict = Body(...),
    user: AuthenticatedUser = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """Update an execution hook. Requires admin."""
    hook = await _get_hook(hook_id, db)
    attrs = body.get("data", {}).get("attributes", {})
    if "name" in attrs:
        new_name = (attrs["name"] or "").strip()
        if not new_name:
            raise HTTPException(status_code=422, detail="Execution hook name is required")
        hook.name = new_name
    if "description" in attrs:
        hook.description = attrs["description"]
    if "hook-point" in attrs:
        execution_hook_service.validate_hook_point(attrs["hook-point"])
        hook.hook_point = attrs["hook-point"]
    if "script" in attrs:
        hook.script = attrs["script"]
    if "enabled" in attrs:
        hook.enabled = bool(attrs["enabled"])
    if "priority" in attrs:
        hook.priority = int(attrs["priority"] or 0)

    try:
        await db.commit()
    except IntegrityError as exc:
        await db.rollback()
        raise HTTPException(
            status_code=409, detail=f"An execution hook named '{hook.name}' already exists"
        ) from exc
    await db.refresh(hook, ["workspace_assignments"])
    return JSONResponse(content={"data": _hook_json(hook)})


@router.delete("/execution-hooks/{hook_id}", status_code=204)
async def delete_hook(
    hook_id: str = Path(...),
    user: AuthenticatedUser = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> None:
    """Delete an execution hook. Requires admin."""
    hook = await _get_hook(hook_id, db)
    await db.delete(hook)
    await db.commit()


@router.post("/execution-hooks/{hook_id}/relationships/workspaces", status_code=204)
async def add_hook_workspaces(
    hook_id: str = Path(...),
    body: dict = Body(...),
    user: AuthenticatedUser = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> None:
    """Associate workspaces with a hook. Requires admin. Idempotent."""
    hook = await _get_hook(hook_id, db)
    for item in body.get("data", []):
        ws_id = (item.get("id") or "").removeprefix("ws-")
        try:
            ws_uuid = uuid.UUID(ws_id)
        except ValueError:
            continue
        ws = await db.get(Workspace, ws_uuid)
        if ws is None:
            continue
        existing = await db.execute(
            select(ExecutionHookWorkspace).where(
                ExecutionHookWorkspace.hook_id == hook.id,
                ExecutionHookWorkspace.workspace_id == ws_uuid,
            )
        )
        if existing.scalar_one_or_none() is None:
            db.add(ExecutionHookWorkspace(hook_id=hook.id, workspace_id=ws_uuid))
    await db.commit()


@router.delete("/execution-hooks/{hook_id}/relationships/workspaces", status_code=204)
async def remove_hook_workspaces(
    hook_id: str = Path(...),
    body: dict = Body(...),
    user: AuthenticatedUser = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> None:
    """Remove workspace associations from a hook. Requires admin."""
    hook = await _get_hook(hook_id, db)
    for item in body.get("data", []):
        ws_id = (item.get("id") or "").removeprefix("ws-")
        try:
            ws_uuid = uuid.UUID(ws_id)
        except ValueError:
            continue
        result = await db.execute(
            select(ExecutionHookWorkspace).where(
                ExecutionHookWorkspace.hook_id == hook.id,
                ExecutionHookWorkspace.workspace_id == ws_uuid,
            )
        )
        assoc = result.scalar_one_or_none()
        if assoc:
            await db.delete(assoc)
    await db.commit()

File: go-terrapod/execution_hooks.go
package terrapod

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
)

// ExecutionHook is a reusable custom-shell step run inside the runner Job at a
// fixed point (pre_init/pre_plan/post_plan/pre_apply/post_apply). A hook reaches
// a workspace only via explicit association (AssignWorkspaceToExecutionHook) —
// there is no global flag. Managing hooks requires platform admin.
type ExecutionHook struct {
	ID             string `json:"id"`
	Name           string `json:"name"`
	Description    string `json:"description,omitempty"`
	HookPoint      string `json:"hook-point"`
	Script         string `json:"script,omitempty"`
	Enabled        bool   `json:"enabled"`
	Priority       int64  `json:"priority"`
	WorkspaceCount int64  `json:"workspace-count"`
	CreatedAt      string `json:"created-at,omitempty"`
	UpdatedAt      string `json:"updated-at,omitempty"`
}

// CreateExecutionHookRequest is the input shape for CreateExecutionHook.
type CreateExecutionHookRequest struct {
	Name        string
	Description string
	HookPoint   string
	Script      string
	Enabled     bool
	Priority    int64
}

// UpdateExecutionHookRequest is the partial-update shape (nil pointer = leave
// unchanged).
type UpdateExecutionHookRequest struct {
	Name        string
	Description *string
	HookPoint   *string
	Script      *string
	Enabled     *bool
	Priority    *int64
}

// CreateExecutionHook creates a new execution hook.
func (c *Client) CreateExecutionHook(ctx context.Context, req CreateExecutionHookRequest) (*ExecutionHook, error) {
	body, err := MarshalResource("execution-hooks", executionHookCreateAttrs(req), nil)
	if err != nil {
		return nil, fmt.Errorf("marshal create execution hook: %w", err)
	}
	data, err := c.Post(ctx, "/api/terrapod/v1/execution-hooks", body)
	if err != nil {
		return nil, err
	}
	return parseExecutionHook(data)
}

// GetExecutionHook reads a hook by id.
func (c *Client) GetExecutionHook(ctx context.Context, id string) (*ExecutionHook, error) {
	data, err := c.Get(ctx, "/api/terrapod/v1/execution-hooks/"+url.PathEscape(id))
	if err != nil {
		return nil, err
	}
	return parseExecutionHook(data)
}

// ListExecutionHooks returns every execution hook.
func (c *Client) ListExecutionHooks(ctx context.Context) ([]ExecutionHook, error) {
	data, err := c.Get(ctx, "/api/terrapod/v1/execution-hooks")
	if err != nil {
		return nil, err
	}
	resources, err := ParseResourceList(data)
	if err != nil {
		return nil, err
	}
	out := make([]ExecutionHook, 0, len(resources))
	for i := range resources {
		out = append(out, *executionHookFromResource(&resources[i]))
	}
	return out, nil
}

// UpdateExecutionHook patches a hook.
func (c *Client) UpdateExecutionHook(ctx context.Context, id string, req UpdateExecutionHookRequest) (*ExecutionHook, error) {
	body, err := MarshalResourceWithID(id, "execution-hooks", executionHookUpdateAttrs(req))
	if err != nil {
		return nil, fmt.Errorf("marshal update execution hook: %w", err)
	}
	data, err := c.Patch(ctx, "/api/terrapod/v1/execution-hooks/"+url.PathEscape(id), body)
	if err != nil {
		return nil, err
	}
	return parseExecutionHook(data)
}

// DeleteExecutionHook removes a hook and all its workspace associations.
func (c *Client) DeleteExecutionHook(ctx context.Context, id string) error {
	return c.Delete(ctx, "/api/terrapod/v1/execution-hooks/"+url.PathEscape(id))
}

// ── Workspace association ────────────────────────────────────────────

// AssignWorkspaceToExecutionHook associates a workspace with a hook so the hook
// runs on that workspace's agent runs.
func (c *Client) AssignWorkspaceToExecutionHook(ctx context.Context, hookID, workspaceID string) error {
	body, err := marshalRelData([]relRef{{ID: workspaceID, Type: "workspaces"}})
	if err != nil {
		return fmt.Errorf("marshal assign body: %w", err)
	}
	_, err = c.Post(ctx,
		fmt.Sprintf("/api/terrapod/v1/execution-hooks/%s/relationships/workspaces", url.PathEscape(hookID)),
		body)
	return err
}

// UnassignWorkspaceFromExecutionHook removes the association. Idempotent.
func (c *Client) UnassignWorkspaceFromExecutionHook(ctx context.Context, hookID, workspaceID string) error {
	body, err := marshalRelData([]relRef{{ID: workspaceID, Type: "workspaces"}})
	if err != nil {
		return fmt.Errorf("marshal unassign body: %w", err)
	}
	return c.DeleteWithBody(ctx,
		fmt.Sprintf("/api/terrapod/v1/execution-hooks/%s/relationships/workspaces", url.PathEscape(hookID)),
		body)
}

// IsWorkspaceAssignedToExecutionHook reports whether the workspace is in the
// hook's `workspaces` relationship. Helper for Terraform Read paths.
func (c *Client) IsWorkspaceAssignedToExecutionHook(ctx context.Context, hookID, workspaceID string) (bool, error) {
	data, err := c.Get(ctx, "/api/terrapod/v1/execution-hooks/"+url.PathEscape(hookID))
	if err != nil {
		return false, err
	}
	res, err := ParseResource(data)
	if err != nil {
		return false, err
	}
	rel, ok := res.Relationships["workspaces"]
	if !ok || len(rel.Data) == 0 || string(rel.Data) == "null" {
		return false, nil
	}
	var items []RelationshipResource
	if err := json.Unmarshal(rel.Data, &items); err != nil {
		return false, nil
	}
	for _, item := range items {
		if item.ID == workspaceID {
			return true, nil
		}
	}
	return false, nil
}

// ── Internal helpers ─────────────────────────────────────────────────

func executionHookCreateAttrs(req CreateExecutionHookRequest) map[string]any {
	attrs := map[string]any{
		"name":       req.Name,
		"hook-point": req.HookPoint,
		"script":     req.Script,
		"enabled":    req.Enabled,
		"priority":   req.Priority,
	}
	if req.Description != "" {
		attrs["description"] = req.Description
	}
	return attrs
}

func executionHookUpdateAttrs(req UpdateExecutionHookRequest) map[string]any {
	attrs := map[string]any{}
	if req.Name != "" {
		attrs["name"] = req.Name
	}
	if req.Description != nil {
		attrs["description"] = *req.Description
	}
	if req.HookPoint != nil {
		attrs["hook-point"] = *req.HookPoint
	}
	if req.Script != nil {
		attrs["script"] = *req.Script
	}
	if req.Enabled != nil {
		attrs["enabled"] = *req.Enabled
	}
	if req.Priority != nil {
		attrs["priority"] = *req.Priority
	}
	return attrs
}

func parseExecutionHook(body []byte) (*ExecutionHook, error) {
	res, err := ParseResource(body)
	if err != nil {
		return nil, fmt.Errorf("parse execution hook response: %w", err)
	}
	return executionHookFromResource(res), nil
}

func executionHookFromResource(res *Resource) *ExecutionHook {
	return &ExecutionHook{
		ID:             res.ID,
		Name:           GetStringAttr(res, "name"),
		Description:    GetStringAttr(res, "description"),
		HookPoint:      GetStringAttr(res, "hook-point"),
		Script:         GetStringAttr(res, "script"),
		Enabled:        GetBoolAttr(res, "enabled"),
		Priority:       GetIntAttr(res, "priority"),
		WorkspaceCount: GetIntAttr(res, "workspace-count"),
		CreatedAt:      GetStringAttr(res, "created-at"),
		UpdatedAt:      GetStringAttr(res, "updated-at"),
	}
}
